libraries/AP_HAL_Linux/CAN.cpp

```diff
@@ -0,0 +1,612 @@
+}
+
+bool LinuxCAN::hasReadyRx() const
+{
+    return !_rx_queue.empty();
+}
+
+void LinuxCAN::poll(bool read, bool write)
+{
+    if (read) {
+        _pollRead();  // Read poll must be executed first because it may decrement _frames_in_socket_tx_queue
+    }
+    if (write) {
+        _pollWrite();
+    }
+}
+
+int16_t LinuxCAN::configureFilters(const uavcan::CanFilterConfig* const filter_configs,
+                              const std::uint16_t num_configs)
+{
+    if (filter_configs == nullptr) {
+        return -1;
+    }
+    _hw_filters_container.clear();
+    _hw_filters_container.resize(num_configs);
+
+    for (unsigned i = 0; i < num_configs; i++) {
+        const uavcan::CanFilterConfig& fc = filter_configs[i];
+        _hw_filters_container[i].can_id   = fc.id   & uavcan::CanFrame::MaskExtID;
+        _hw_filters_container[i].can_mask = fc.mask & uavcan::CanFrame::MaskExtID;
+        if (fc.id & uavcan::CanFrame::FlagEFF) {
+            _hw_filters_container[i].can_id |= CAN_EFF_FLAG;
+        }
+        if (fc.id & uavcan::CanFrame::FlagRTR) {
+            _hw_filters_container[i].can_id |= CAN_RTR_FLAG;
+        }
+        if (fc.mask & uavcan::CanFrame::FlagEFF) {
+            _hw_filters_container[i].can_mask |= CAN_EFF_FLAG;
+        }
+        if (fc.mask & uavcan::CanFrame::FlagRTR) {
+            _hw_filters_container[i].can_mask |= CAN_RTR_FLAG;
+        }
+    }
+
+    return 0;
+}
+
+/**
+ * SocketCAN emulates the CAN filters in software, so the number of filters is virtually unlimited.
+ * This method returns a constant value.
+ */
+static constexpr unsigned NumFilters = 8;
+uint16_t LinuxCAN::getNumFilters() const { return NumFilters; }
+
+uint64_t LinuxCAN::getErrorCount() const
+{
+    uint64_t ec = 0;
+    for (auto& kv : _errors) { ec += kv.second; }
+    return ec;
+}
+
+void LinuxCAN::_pollWrite()
+{
+    while (hasReadyTx()) {
+        const TxItem tx = _tx_queue.top();
+
+        if (tx.deadline >= getMonotonic()) {
+            const int res = _write(tx.frame);
+            if (res == 1) {                   // Transmitted successfully
+                _incrementNumFramesInSocketTxQueue();
+                if (tx.flags & uavcan::CanIOFlagLoopback) {
+                    _pending_loopback_ids.insert(tx.frame.id);
+                }
+            } else if (res == 0) {            // Not transmitted, nor is it an error
+                break;                        // Leaving the loop, the frame remains enqueued for the next retry
+            } else {                          // Transmission error
+                _registerError(SocketCanError::SocketWriteFailure);
+            }
+        } else {
+            _registerError(SocketCanError::TxTimeout);
+        }
+
+        // Removing the frame from the queue even if transmission failed
+        _tx_queue.pop();
+    }
+}
+
+void LinuxCAN::_pollRead()
+{
+    while (true)
```

Comment:
can we get a counter here to only loop to a maximum number of iterations? This avoids situations in which we would be stuck processing packets here.